## Check for Duplicates

In [1]:
import pandas as pd


# Define the dataset path
dataset_path = '/content/drive/MyDrive/Diamond Dataset/Datasets/Final Datatset.csv'

# Load the dataset
try:
    df = pd.read_csv(dataset_path)
    print(f"Dataset loaded successfully from: {dataset_path}")
    print(f"Shape of the dataset: {df.shape}")
except FileNotFoundError:
    print(f"Error: The file '{dataset_path}' was not found. Please ensure the path is correct and the file is accessible.")
    df = None
except Exception as e:
    print(f"An error occurred while loading the dataset: {e}")
    df = None

if df is not None:
    # Check for duplicate rows including all occurrences
    all_duplicates = df[df.duplicated(keep=False)]
    num_duplicates_all_occurrences = len(all_duplicates)

    if num_duplicates_all_occurrences > 0:
        print(f"\nNumber of duplicate rows (including all occurrences): {num_duplicates_all_occurrences}")
        print("Here are the first 10 duplicate rows (including all occurrences):")
        # Display more rows to show multiple occurrences of the same duplicate set
        print(all_duplicates.head(10))
    else:
        print("\nNo duplicate rows found.")

    # Acknowledge the target variable
    target_variable = 'price'
    print(f"\nTarget variable identified: '{target_variable}'")


Error: The file '/content/drive/MyDrive/Diamond Dataset/Datasets/Final Datatset.csv' was not found. Please ensure the path is correct and the file is accessible.


## Removing Duplicates

In [ ]:
# Calculate initial number of rows
initial_rows = df.shape[0]

# Remove duplicate rows, keeping the first occurrence
df_cleaned = df.drop_duplicates(keep='first')

# Calculate the number of rows deleted
deleted_rows = initial_rows - df_cleaned.shape[0]

print(f"Number of rows before removing duplicates: {initial_rows}")
print(f"Number of rows after removing duplicates: {df_cleaned.shape[0]}")
print(f"Number of duplicate rows deleted: {deleted_rows}")

# Update the main DataFrame to the cleaned version
df = df_cleaned.copy()


Number of rows before removing duplicates: 80907
Number of rows after removing duplicates: 80727
Number of duplicate rows deleted: 180


## Test Train Split

In [ ]:
# Combine X_train and y_train into a single training DataFrame
train_df = pd.concat([X_train.reset_index(drop=True), y_train.reset_index(drop=True)], axis=1)

# Combine X_test and y_test into a single testing DataFrame
test_df = pd.concat([X_test.reset_index(drop=True), y_test.reset_index(drop=True)], axis=1)

print(f"Shape of the training DataFrame (train_df): {train_df.shape}")
print(f"Shape of the testing DataFrame (test_df): {test_df.shape}")

print("\nFirst 5 rows of train_df:")
display(train_df.head())

print("\nFirst 5 rows of test_df:")
display(test_df.head())


Shape of the training DataFrame (train_df): (64581, 11)
Shape of the testing DataFrame (test_df): (16146, 11)

First 5 rows of train_df:


,Type,Carat,Cut,Color,Clarity,Depth,Table,X,Y,Z,Price
0,Diamond,0.80,Very Good,E,VVS2,62.8,57.0,5.93,5.95,3.73,4893.00
1,Zirconia,0.31,Ideal,H,VS1,61.7,57.0,4.34,4.38,2.69,5.98
2,Diamond,1.21,Very Good,I,VVS2,61.1,57.0,6.83,6.85,4.18,6653.00
3,Diamond,0.31,Ideal,H,VVS1,61.6,55.0,4.37,4.40,2.70,687.00
4,Diamond,1.00,Very Good,F,VS1,58.1,62.0,6.57,6.62,3.83,6098.00



First 5 rows of test_df:


,Type,Carat,Cut,Color,Clarity,Depth,Table,X,Y,Z,Price
0,Diamond,0.30,Ideal,G,IF,62.3,55.0,4.30,4.33,2.69,863.00
1,Zirconia,1.17,Ideal,J,VS2,62.1,56.0,6.80,6.75,4.21,50.81
2,Diamond,1.25,Ideal,J,SI1,62.7,57.0,6.86,6.88,4.31,4764.00
3,Zirconia,0.40,Premium,E,SI1,62.8,59.0,4.73,4.70,2.96,9.98
4,Diamond,1.26,Premium,E,SI2,60.3,60.0,7.00,6.94,4.20,5058.00


## Missing Value Test

In [ ]:
print("Checking for missing values in the Training Data (train_df):")
if 'train_df' in locals() and train_df is not None:
    train_missing_values = train_df.isnull().sum()
    train_total_missing_values = train_missing_values.sum()

    print("Missing values per column in train_df:")
    print(train_missing_values[train_missing_values > 0])

    if train_total_missing_values == 0:
        print("\nNo missing values found in the training dataset.")
    else:
        print(f"\nTotal number of missing values in the training dataset: {train_total_missing_values}")
else:
    print("train_df is not available. Cannot check for missing values.")

print("\n---\n")

print("Checking for missing values in the Testing Data (test_df):")
if 'test_df' in locals() and test_df is not None:
    test_missing_values = test_df.isnull().sum()
    test_total_missing_values = test_missing_values.sum()

    print("Missing values per column in test_df:")
    print(test_missing_values[test_missing_values > 0])

    if test_total_missing_values == 0:
        print("\nNo missing values found in the testing dataset.")
    else:
        print(f"\nTotal number of missing values in the testing dataset: {test_total_missing_values}")
else:
    print("test_df is not available. Cannot check for missing values.")


Checking for missing values in the Training Data (train_df):
Missing values per column in train_df:
Depth    567
dtype: int64

Total number of missing values in the training dataset: 567

---

Checking for missing values in the Testing Data (test_df):
Missing values per column in test_df:
Depth    130
dtype: int64

Total number of missing values in the testing dataset: 130


## Missing value Imputation
Impute missing 'Depth' values in both the `train_df` and `test_df` DataFrames using a regression model trained on the non-missing 'Depth' data from `train_df`. The imputation process should involve separating `train_df` into subsets for training and prediction, preprocessing numerical features with `StandardScaler` and categorical features with `OneHotEncoder`, training a regression model (e.g., RandomForestRegressor) on the `train_df` subset with known 'Depth' values, and then using this trained model to predict and fill the missing 'Depth' values in both `train_df` and `test_df`. Finally, verify that no 'Depth' missing values remain in either DataFrame.

## Prepare Data for Imputation Model

### Subtask:
Separate the `train_df` into two subsets: one with known 'Depth' values (for training the imputation model) and one with missing 'Depth' values (for prediction). Define features and target for the imputation model.


**Reasoning**:
To prepare for missing value imputation, I will separate the training data into two dataframes: one for rows with existing 'Depth' values (for training the imputation model) and another for rows with missing 'Depth' values (for predicting 'Depth'). Then I will define the features and target for the imputation model.



In [ ]:
print("Splitting train_df into subsets for Depth imputation...")

# 1. Create two new DataFrames from train_df
train_with_depth = train_df[train_df['Depth'].notna()].copy()
train_without_depth = train_df[train_df['Depth'].isna()].copy()

print(f"Shape of train_with_depth (rows with known 'Depth'): {train_with_depth.shape}")
print(f"Shape of train_without_depth (rows with missing 'Depth'): {train_without_depth.shape}")

# 2. Define features (X_impute) and target (y_impute) for the imputation model
# Exclude 'Depth' (target) and 'Price' (actual target variable of the main problem) from features
imputation_features = [col for col in train_df.columns if col not in ['Depth', 'Price']]
X_impute = train_with_depth[imputation_features]
y_impute = train_with_depth['Depth']

print(f"\nShape of X_impute (features for imputation model training): {X_impute.shape}")
print(f"Shape of y_impute (target 'Depth' for imputation model training): {y_impute.shape}")

# 3. Define the features (X_predict_depth) from the train_without_depth DataFrame
X_predict_depth = train_without_depth[imputation_features]

print(f"Shape of X_predict_depth (features for predicting missing 'Depth'): {X_predict_depth.shape}")

print("\nFirst 5 rows of X_impute:")
display(X_impute.head())
print("\nFirst 5 rows of y_impute:")
display(y_impute.head())
print("\nFirst 5 rows of X_predict_depth:")
display(X_predict_depth.head())

Splitting train_df into subsets for Depth imputation...
Shape of train_with_depth (rows with known 'Depth'): (64014, 11)
Shape of train_without_depth (rows with missing 'Depth'): (567, 11)

Shape of X_impute (features for imputation model training): (64014, 9)
Shape of y_impute (target 'Depth' for imputation model training): (64014,)
Shape of X_predict_depth (features for predicting missing 'Depth'): (567, 9)

First 5 rows of X_impute:


,Type,Carat,Cut,Color,Clarity,Table,X,Y,Z
0,Diamond,0.80,Very Good,E,VVS2,57.0,5.93,5.95,3.73
1,Zirconia,0.31,Ideal,H,VS1,57.0,4.34,4.38,2.69
2,Diamond,1.21,Very Good,I,VVS2,57.0,6.83,6.85,4.18
3,Diamond,0.31,Ideal,H,VVS1,55.0,4.37,4.40,2.70
4,Diamond,1.00,Very Good,F,VS1,62.0,6.57,6.62,3.83



First 5 rows of y_impute:


,Depth
0,62.8
1,61.7
2,61.1
3,61.6
4,58.1



First 5 rows of X_predict_depth:


,Type,Carat,Cut,Color,Clarity,Table,X,Y,Z
240,Zirconia,0.30,Premium,H,VS2,60.0,4.26,4.31,2.65
264,Zirconia,0.54,Very Good,F,VS2,60.0,5.34,5.39,3.12
362,Zirconia,2.24,Premium,J,SI2,59.0,8.37,8.33,5.19
441,Zirconia,0.33,Ideal,D,VS2,55.0,4.47,4.50,2.73
645,Zirconia,0.32,Ideal,E,VVS1,57.0,4.42,4.45,2.72


**Reasoning**:
Now that the data for imputation has been separated into features and target, the next step is to preprocess these features. Numerical features will be scaled using `StandardScaler`, and categorical features will be encoded using `OneHotEncoder`. This is essential for preparing the data for a regression model.



In [ ]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.pipeline import Pipeline

print("Identifying numerical and categorical features for preprocessing...")

# Identify numerical and categorical features for the imputation model
# Exclude 'Price' as it is the main target variable and 'Depth' is the target for imputation
numerical_features = X_impute.select_dtypes(include=['number']).columns.tolist()
categorical_features = X_impute.select_dtypes(include=['object']).columns.tolist()

print(f"Numerical features: {numerical_features}")
print(f"Categorical features: {categorical_features}")

# Create a preprocessor using ColumnTransformer
# Numerical features: scale using StandardScaler
# Categorical features: encode using OneHotEncoder
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numerical_features),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_features)
    ])

print("Fitting preprocessor on X_impute and transforming X_impute and X_predict_depth...")

# Fit the preprocessor on X_impute and transform it
X_impute_processed = preprocessor.fit_transform(X_impute)

# Transform X_predict_depth using the *fitted* preprocessor
X_predict_depth_processed = preprocessor.transform(X_predict_depth)

print(f"Shape of X_impute_processed: {X_impute_processed.shape}")
print(f"Shape of X_predict_depth_processed: {X_predict_depth_processed.shape}")

Identifying numerical and categorical features for preprocessing...
Numerical features: ['Carat', 'Table', 'X', 'Y', 'Z']
Categorical features: ['Type', 'Cut', 'Color', 'Clarity']
Fitting preprocessor on X_impute and transforming X_impute and X_predict_depth...
Shape of X_impute_processed: (64014, 27)
Shape of X_predict_depth_processed: (567, 27)


**Reasoning**:
With the features preprocessed, the next step is to train a regression model to predict the 'Depth' values. I will use a `RandomForestRegressor` for this purpose, fitting it on the processed features and known 'Depth' values.



In [ ]:
from sklearn.ensemble import RandomForestRegressor

print("Training RandomForestRegressor for Depth imputation...")

# Initialize the RandomForestRegressor model
# Using default parameters for now, can be tuned if needed
imputation_model = RandomForestRegressor(random_state=42, n_jobs=-1)

# Train the model on the processed features and known 'Depth' values
imputation_model.fit(X_impute_processed, y_impute)

print("RandomForestRegressor training complete.")

Training RandomForestRegressor for Depth imputation...
RandomForestRegressor training complete.


**Reasoning**:
With the imputation model trained, the next step is to use it to predict the missing 'Depth' values in the `train_without_depth` DataFrame. These predicted values will then be used to fill the `Depth` column in the original `train_df` where values were missing.



In [ ]:
print("Predicting missing 'Depth' values for train_df...")

# Predict the missing 'Depth' values using the trained model
predicted_depth_train = imputation_model.predict(X_predict_depth_processed)

# Fill the missing 'Depth' values in the original train_df
train_df.loc[train_df['Depth'].isna(), 'Depth'] = predicted_depth_train

print("Missing 'Depth' values in train_df have been imputed.")
print(f"Number of missing 'Depth' values in train_df after imputation: {train_df['Depth'].isnull().sum()}")

Predicting missing 'Depth' values for train_df...
Missing 'Depth' values in train_df have been imputed.
Number of missing 'Depth' values in train_df after imputation: 0


**Reasoning**:
The missing 'Depth' values in `train_df` have been imputed. Now, I will identify the rows with missing 'Depth' in the `test_df` to prepare for their imputation using the same trained model.



In [ ]:
print("Splitting test_df into subsets for Depth imputation...")

# Create two new DataFrames from test_df
test_with_depth = test_df[test_df['Depth'].notna()].copy()
test_without_depth = test_df[test_df['Depth'].isna()].copy()

print(f"Shape of test_with_depth (rows with known 'Depth'): {test_with_depth.shape}")
print(f"Shape of test_without_depth (rows with missing 'Depth'): {test_without_depth.shape}")

# Define the features (X_predict_depth_test) from the test_without_depth DataFrame
# Use the same imputation_features identified for the training data
X_predict_depth_test = test_without_depth[imputation_features]

print(f"Shape of X_predict_depth_test (features for predicting missing 'Depth' in test_df): {X_predict_depth_test.shape}")

print("\nFirst 5 rows of X_predict_depth_test:")
display(X_predict_depth_test.head())

Splitting test_df into subsets for Depth imputation...
Shape of test_with_depth (rows with known 'Depth'): (16016, 11)
Shape of test_without_depth (rows with missing 'Depth'): (130, 11)
Shape of X_predict_depth_test (features for predicting missing 'Depth' in test_df): (130, 9)

First 5 rows of X_predict_depth_test:


,Type,Carat,Cut,Color,Clarity,Table,X,Y,Z
254,Zirconia,1.04,Premium,H,SI1,59.0,6.59,6.55,3.97
425,Zirconia,0.50,Good,E,VVS2,62.0,5.04,5.09,3.12
503,Zirconia,0.70,Fair,E,VS2,61.0,5.62,5.47,3.64
524,Zirconia,1.22,Ideal,G,VS2,55.0,6.83,6.87,4.25
657,Zirconia,1.47,Fair,H,I1,59.0,6.98,6.94,4.64


**Reasoning**:
Now that the `test_df` has been split and the features for predicting missing 'Depth' (`X_predict_depth_test`) have been identified, I need to preprocess these features using the same `preprocessor` that was fitted on the training data. This ensures consistency in scaling and encoding before making predictions.



In [ ]:
print("Transforming X_predict_depth_test using the fitted preprocessor...")

# Transform X_predict_depth_test using the *already fitted* preprocessor
X_predict_depth_test_processed = preprocessor.transform(X_predict_depth_test)

print(f"Shape of X_predict_depth_test_processed: {X_predict_depth_test_processed.shape}")

Transforming X_predict_depth_test using the fitted preprocessor...
Shape of X_predict_depth_test_processed: (130, 27)


**Reasoning**:
With the test data features now preprocessed, I will use the previously trained `imputation_model` to predict the missing 'Depth' values for `test_df`. Then, I will fill the missing 'Depth' values in the original `test_df` with these predictions and verify that no missing values remain.



In [ ]:
print("Predicting missing 'Depth' values for test_df...")

# Predict the missing 'Depth' values using the trained model
predicted_depth_test = imputation_model.predict(X_predict_depth_test_processed)

# Fill the missing 'Depth' values in the original test_df
test_df.loc[test_df['Depth'].isna(), 'Depth'] = predicted_depth_test

print("Missing 'Depth' values in test_df have been imputed.")
print(f"Number of missing 'Depth' values in test_df after imputation: {test_df['Depth'].isnull().sum()}")

# Final check for missing values in both train_df and test_df
print(f"\nFinal check: Missing values in train_df['Depth']: {train_df['Depth'].isnull().sum()}")
print(f"Final check: Missing values in test_df['Depth']: {test_df['Depth'].isnull().sum()}")

Predicting missing 'Depth' values for test_df...
Missing 'Depth' values in test_df have been imputed.
Number of missing 'Depth' values in test_df after imputation: 0

Final check: Missing values in train_df['Depth']: 0
Final check: Missing values in test_df['Depth']: 0


## Summary:

### Q&A
The task of imputing missing 'Depth' values in both the `train_df` and `test_df` using a regression model, and subsequently verifying that no 'Depth' missing values remain, has been successfully completed.

### Data Analysis Key Findings
*   Initially, `train_df` contained 567 rows with missing 'Depth' values, while `test_df` had 130 rows with missing 'Depth' values.
*   For the imputation model, `train_df` was split into a training subset of 64014 rows with known 'Depth' and a prediction subset of 567 rows with missing 'Depth'. Similarly, `test_df` was split into 16016 rows with known 'Depth' and 130 rows with missing 'Depth'.
*   The imputation model utilized 9 features (e.g., 'Carat', 'Table', 'Type', 'Cut', 'Color', 'Clarity', 'X', 'Y', 'Z'), excluding 'Depth' and 'Price'.
*   Preprocessing involved scaling numerical features with `StandardScaler` and encoding categorical features with `OneHotEncoder`, transforming the feature space into 27 columns for the model.
*   A `RandomForestRegressor` was successfully trained on the preprocessed data from `train_df` with known 'Depth' values.
*   The trained `RandomForestRegressor` was used to predict and fill the 567 missing 'Depth' values in `train_df` and the 130 missing 'Depth' values in `test_df`.
*   Post-imputation, both `train_df` and `test_df` now have 0 missing values in their respective 'Depth' columns.

### Insights or Next Steps
*   The regression-based imputation strategy effectively handled all missing 'Depth' values, allowing for comprehensive data utilization in subsequent modeling phases.
*   Consider evaluating the performance of the imputation model (e.g., through cross-validation on the known 'Depth' subset) or comparing it with other imputation methods to optimize the accuracy of the imputed 'Depth' values.


## Model use for Imputation

In [ ]:
import joblib
import os

# Define the path to save the model in Google Drive
# Make sure the directory exists or create it if not
save_dir = '/content/drive/MyDrive/Diamond Dataset/models'
os.makedirs(save_dir, exist_ok=True)
model_path = os.path.join(save_dir, 'imputation_random_forest_regressor.joblib')

# Save the trained imputation model
joblib.dump(imputation_model, model_path)

print(f"The imputation model has been successfully saved to: {model_path}")


The imputation model has been successfully saved to: /content/drive/MyDrive/Diamond Dataset/models/imputation_random_forest_regressor.joblib


## Download Datasets

In [ ]:
import os

# Define the directory to save the datasets in Google Drive
save_dataset_dir = '/content/drive/MyDrive/Diamond Dataset/Datasets'
os.makedirs(save_dataset_dir, exist_ok=True)

# Define paths for the train and test CSV files
train_csv_path = os.path.join(save_dataset_dir, 'train_df.csv')
test_csv_path = os.path.join(save_dataset_dir, 'test_df.csv')

# Save the DataFrames to CSV
if 'train_df' in locals() and train_df is not None:
    train_df.to_csv(train_csv_path, index=False)
    print(f"'train_df' saved successfully to: {train_csv_path}")
else:
    print("'train_df' is not available. Cannot save to CSV.")

if 'test_df' in locals() and test_df is not None:
    test_df.to_csv(test_csv_path, index=False)
    print(f"'test_df' saved successfully to: {test_csv_path}")
else:
    print("'test_df' is not available. Cannot save to CSV.")


'train_df' saved successfully to: /content/drive/MyDrive/Diamond Dataset/datasets/train_df.csv
'test_df' saved successfully to: /content/drive/MyDrive/Diamond Dataset/datasets/test_df.csv
